packages/flutter_tools/lib/src/android/android_device.dart

```diff
@@ -90,8 +101,17 @@ class AndroidDevice extends Device {
   @override
   Future<bool> get isLocalEmulator async {
     if (_isLocalEmulator == null) {
-      final String characteristics = await _getProperty('ro.build.characteristics');
-      _isLocalEmulator = characteristics != null && characteristics.contains('emulator');
+      final String hardware = await _getProperty('ro.hardware');
+      printTrace('ro.hardware = $hardware');
+      if (_knownHardware.containsKey(hardware)) {
+        // Look for known hardware models.
+        _isLocalEmulator = _knownHardware[hardware] == _HardwareType.emulator;
+      } else {
+        // Fall back to a best-effort heuristic-based approach.
+        final String characteristics = await _getProperty('ro.build.characteristics');
+        printTrace('ro.build.characteristics = $characteristics');
+        _isLocalEmulator = characteristics != null && characteristics.contains('emulator');
```

Comment:
That's why I filed and referenced #10248, which tracks that problem.  There are other places where we use the `isLocalEmulator` property though (like in analytics), so this was still worthwhile.